common/recommender/Baseline.js

```diff
@@ -15,27 +15,134 @@ class Baseline {
   constructor(history, options = {}) {
     this.domainCounts = history.reduce(this.countDomainOccurrences, new Map());
     this.options = options;
+    // Features that are extracted from URLs and need normalization.
+    // Key 0 holds the min, key 1 holds max, using arrays for brevity.
+    this.normalizeFeatures = {
+      idf: [1, 0],
+      imageCount: [1, 0],
+      description: [1, 0],
+      pathLength: [1, 0],
+      largestImage: [1, 0],
+      queryLength: [1, 0]
+    };
   }
 
-  scoreEntry(entry) {
+  extractFeatures(entry) {
     const urlObj = URL(entry.url);
-    const host = URL(entry.url).host;
+    const host = urlObj.host;
+    const occurences = this.domainCounts.get(host) || 1;
+
+    const age = this.normalizeTimestamp(entry.lastVisitDate);
     const tf = Math.max(entry.visitCount, 1);
-    const idf = Math.log(this.domainCounts.size / Math.max(1, this.domainCounts.get(host)));
+    const idf = Math.log(this.domainCounts.size / occurences);
     const imageCount = entry.images ? entry.images.length : 0;
-    const isBookmarked = entry.bookmarkId > 0 ? 1 : 0;
-    const hasDescription = entry.title !== entry.description ? 1 : 0;
-
-    const score = this.decay(tf * idf, // Score
-      // Features: Age in hours, number of visits to url, url query length, number of images, is a bookmark,
-      //           has description.
-      [this.normalizeTimestamp(entry.lastVisitDate), entry.visitCount, urlObj.query.length,
-       imageCount, isBookmarked, hasDescription],
+    const isBookmarked = entry.bookmarkId || 0;
+    const description = this.extractDescriptionLength(entry);
+    const pathLength = urlObj.pathname.split("/").filter(e => e.length).length;
+    const largestImage = this.extractLargestImage(entry.images);
+    const queryLength = urlObj.query.length;
+
+    const features = {age, tf, idf, imageCount, isBookmarked, description, pathLength, queryLength, largestImage};
+    this.updateFeatureMinMax(features);
+
+    return Object.assign({}, entry, {features, host});
+  }
+
+  normalize(features) {
+    let delta;
+    const normalizeKeys = Object.keys(this.normalizeFeatures);
+
+    normalizeKeys.forEach(key => {
```

Comment:
this would also be a lot more readable with max/min instead of array indices, i.e.
```js
  normalize(features) {
    Object.keys(this.normalizeFeatures).forEach(key => {
      const {min, max} = this.normalizeFeatures[key];
      if (max !== min) { // No division by 0.
        let delta = max - min;
        features[key] = (features[key] - max) / delta;
      }
    });
  }
```